connector/logs_router_test.go

```diff
@@ -0,0 +1,156 @@
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+	"go.opentelemetry.io/collector/internal/testdata"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type mutatingLogsSink struct {
+	*consumertest.LogsSink
+}
+
+func (mts *mutatingLogsSink) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
+}
+
+func TestLogsMultiplexing(t *testing.T) {
+	var max = 20
+	for numIDs := 1; numIDs < max; numIDs++ {
+		for numCons := 1; numCons < max; numCons++ {
+			for numLogs := 1; numLogs < max; numLogs++ {
+				t.Run(
+					fmt.Sprintf("%d-ids/%d-cons/%d-logs", numIDs, numCons, numLogs),
+					fuzzLogs(numIDs, numCons, numLogs),
+				)
+			}
+		}
+	}
+}
+
+func fuzzLogs(numIDs, numCons, numLogs int) func(*testing.T) {
+	return func(t *testing.T) {
+		allIDs := make([]component.ID, 0, numCons)
+		allCons := make([]consumer.Logs, 0, numCons)
+		allConsMap := make(map[component.ID]consumer.Logs)
+
+		// If any consumer is mutating, the router must report mutating
+		for i := 0; i < numCons; i++ {
+			allIDs = append(allIDs, component.NewIDWithName("sink", strconv.Itoa(numCons)))
+			// Random chance for each consumer to be mutating
+			if (numCons+numLogs+i)%4 == 0 {
+				allCons = append(allCons, &mutatingLogsSink{LogsSink: new(consumertest.LogsSink)})
+			} else {
+				allCons = append(allCons, new(consumertest.LogsSink))
+			}
+			allConsMap[allIDs[i]] = allCons[i]
+		}
+
+		r := NewLogsRouter(allConsMap)
+		ld := testdata.GenerateLogs(1)
+
+		// Keep track of how many logs each consumer should receive.
+		// This will be validated after every call to RouteLogs.
+		expected := make(map[component.ID]int, numCons)
+
+		for i := 0; i < numLogs; i++ {
+			// Build a random set of ids (no duplicates)
+			randCons := make(map[component.ID]bool, numIDs)
+			for j := 0; j < numIDs; j++ {
+				// This number should be pretty random and less than numCons
+				conNum := (numCons + numIDs + i + j) % numCons
+				randCons[allIDs[conNum]] = true
+			}
+
+			// Convert to slice, update expectations
+			conIDs := make([]component.ID, 0, len(randCons))
+			for id := range randCons {
+				conIDs = append(conIDs, id)
+				expected[id]++
+			}
+
+			// Route to list of consumers
+			fanout, err := r.Consumer(conIDs...)
+			assert.NoError(t, err)
+			assert.NoError(t, fanout.ConsumeLogs(context.Background(), ld))
+
+			// Validate expectations for all consumers
+			for id := range expected {
+				logs := []plog.Logs{}
+				switch con := allConsMap[id].(type) {
+				case *consumertest.LogsSink:
+					logs = con.AllLogs()
+				case *mutatingLogsSink:
+					logs = con.AllLogs()
+				}
+				assert.Len(t, logs, expected[id])
+				for n := 0; n < len(logs); n++ {
+					assert.EqualValues(t, ld, logs[n])
+				}
+			}
+		}
+	}
+}
+
+func TestLogsGetConsumers(t *testing.T) {
```

Comment:
nit: the test names are inconsistent between between data types. Should be `Test[Logs|Metrics|Traces]RouterConsumer`?

Reply:
Done.